python_modules/dagster/dagster/_scheduler/stale.py

```diff
@@ -0,0 +1,57 @@
+from typing import Dict, List, Mapping, Optional, Sequence,
+from dagster._core.definitions.events import AssetKey
+from dagster._core.definitions.logical_version import DEFAULT_LOGICAL_VERSION, CachingProjectedLogicalVersionResolver, LogicalVersion, extract_logical_version_from_entry
+from dagster._core.definitions.run_request import RunRequest
+from dagster._core.host_representation.external import ExternalRepository
+from dagster._core.host_representation.external_data import ExternalAssetNode
+from dagster._core.instance import DagsterInstance
+from dagster._core.workspace.context import WorkspaceProcessContext, WorkspaceRequestContext
+
+import dagster._check as check
+
+def resolve_asset_selection(context: WorkspaceProcessContext, run_request: RunRequest) -> Optional[Sequence[AssetKey]]:
+    if run_request.stale_only:
+        asset_selection = check.not_none(run_request.asset_selection)
+        stale_keys: List[AssetKey] = []
+        request_context = context.create_request_context()
+        repositories = get_repositories(request_context)
+        key_to_node_map = get_asset_nodes_by_asset_key(repositories)
```

Comment:
I recently introduced `ExternalAssetGraph.from_workspace` for building a representation of the asset graph from a `WorkspaceRequestContext`. If we added code versions to `AssetGraph`, I think `CachingProjectedLogicalVersionResolver` might be able to use it instead of `key_to_node_map`.

An additional advantage of that is that, when we incorporate projected versions into asset reconciliation, we'd be able to pass in the `AssetGraph` that the reconciliation sensor has on hand.

Reply:
Not advocating for that refactor before merging this if it would risk making this slip from the next release.